content/4_Gold/Ternary_Search.mdx

```diff
@@ -0,0 +1,145 @@
+---
+id: ternary-search
+title: Ternary Search
+author: Chongtian Ma
+prerequisites:
+  - /silver/binary-search
+description: Searching for a peak in a unimodal function.
+frequency: 1
+---
+<Resources>
+	<Resource
+		source="CP-Algorithms"
+		title="Ternary Search"
+		url="https://cp-algorithms.com/num_methods/ternary_search.html"
+		starred
+	/>
+</Resources>
+
+In calculus terms, ternary search is the equivalent of binary searching for the zero on the
+derivative function. The most basic function can be represented by the standard
+form of a parabola: $ax^2+bx+c$. Depending on if $a$ is positive or negative,
+you are searching for the local minimum or maximum of the function,
+respectively. Note that the function must be unimodal (see the cp-algorithms resource for a definition)!
+
+Unlike binary search where a checking function $f(x)$ only needs to return true
+or false, we need $f(x)$ to output a real number. Between the bounds of $[l, r]$,
+we need to compare the integer output of $f(m_1)$ with $f(m_2)$ where
+$l < m_1 < m_2 < r$ to determine how $l$ and $r$ should be moved.
+
+For example, if the function has a local maximum between $[l, r]$ and
+$f(m_1) < f(m_2)$, then we know the local maximum cannot be between $[l, m_1]$,
+so we continue searching in $[m_1, r]$. For a mini-proof, consider the following
+cases:
+
+- Let's denote the local maximum as $g$. If $m_1$ and $m_2$ are less than $g$,
+  this means that $f(m_1)$ and $f(m_2)$ are both on the increasing side of the
+  function. Therefore, we know $m_2 < g < r$, so we must keep searching between
+  $[m_2, r]$.
+
+- If $m_1 < g$ and $m_2 > g$, then $m_1 < g < m_2$. Therefore, we must keep
+  searching betwen $[m_1, m_2]$.
+
+- Note that $g < m_1 < m_2$ can never happen, becuase the function had to be
+  strictly decreasing after $g$.
+
+Since $[m_1, m_2] \cup [m_2, r] = [m_1, r]$, we cover all cases if we set $l$ to
+$m_1$. A similar proof can be given for $f(m_1) \geq f(m_2)$ or if we are
+searching for a local minimum.
+
+It is usually optimal to set $m_1$ and $m_2$ in a way so that we can maximize the amount of distance we eliminate per
```

Comment:
This is not true when working with real numbers though? in that case you want $m_1$ and $m_2$ to not be too close together, otherwise you risk $f(m_1)$ being incorrectly compared to $f(m_2)$  due to floating point imprecision.